RFCs/rfc-0008-wallet-provider-api.md

```diff
@@ -0,0 +1,387 @@
+
+The "accounts available to the Provider" change when the return value of `mina_accounts` changes.
+
+### Rationale
+The purpose of a Provider is to _provide_ a consumer with access to Mina. In general, a Provider must enable a Mina web application to do two things:
+- Query Mina & the wallet with requests
+- Respond to state changes in the Provider's Mina chain, Client, and Wallet
+
+The Provider API specification consists of a single method and five events. The request method and the message event alone, are sufficient to implement a complete Provider. They are designed to make arbitrary RPC requests and communicate arbitrary messages, respectively.
+
+The remaining four events can be separated into two categories:
+
+- Changes to the Provider's ability to make RPC requests
+    - `connect`
+    - `disconnect`
+- Common Client and/or Wallet state changes that any non-trivial application must handle
+    - `chainChanged`
+    - `accountsChanged`
+
+### Security Considerations
+
+The Provider is intended to pass messages between a Mina Client and a Mina application. It is not responsible for private key or account management; it merely processes RPC messages and emits events. Consequently, account security and user privacy need to be implemented in middlewares between the Provider and its Mina Client. In practice, we call these middleware applications "Wallets," and they usually manage the user's private keys and accounts. The Provider can be thought of as an extension of the Wallet, exposed in an untrusted environment, under the control of some third party (e.g. a website).
+
+#### Handling Adversarial Behavior
+
+Since it is a JavaScript object, consumers can generally perform arbitrary operations on the Provider, and all its properties can be read or overwritten. Therefore, it is best to treat the Provider object as though it is controlled by an adversary. It is paramount that the Provider implementer protects the user, Wallet, and Client by ensuring that:
+
+- The Provider does not contain any private user data.
+- The Provider and Wallet programs are isolated from each other.
+- The Wallet and/or Client rate-limit requests from the Provider.
+- The Wallet and/or Client validate all data sent from the Provider.
+
+### Provider RPC Methods
+This section lists the standard RPC methods that can be supported by the Provider, along with their return types. Each method follows the `mina_{methodName}` naming convention.
+
+#### mina_accounts
+Returns a list of addresses owned by client.
+
+##### Parameters
+None
+
+##### Returns
+An array of hexadecimals as strings representing the addresses owned by the client.
+
+```typescript
+Provider.request({ 
+  method: 'mina_accounts'
+}): Promise<string[]>;
+```
+
+#### mina_chainId
+Returns the chain ID of the current network.
+
+##### Parameters
+None
+
+##### Returns
+A hexadecimal of the current chain ID.
```

Comment:
Is it possible to add a chain name here, for chain with the same chainId but different GQL URLs ?

Reply:
It can do, but is that to make it easier for a zkApp to understand information about the chain? The `mina_chainId` method serves to _only_ provide the current network's chain id -- if there is another method that you think is important to have such as `mina_chainInformation` what would you believe would be an appropriate return object/type?

Reply:
Yes, because some users need to add different proxies for the same network, which can be distinguished by name
This is the structure we return to zkApp now, mainly used to distinguish the network actually used by zkApp
```ts
type ChainInfoArgs = {
  chainId: string;
  name: string;
};
```

Reply:
@lvshaoping007 I'm curious to understand why one would introduce different names for the same network. What is the purpose of having different proxies for the same network?

Reply:
I've added the `mina_chainInformation` method.